### PR TITLE
[FLINK-26588][docs]Translate the new CAST documentation to Chinese

### PR DESCRIPTION
**Pull request by ChengkaiYang2022**
## What is the purpose of the change

Translate the new CAST documentation to Chinese

## Brief change log

  -  Adjust some lines in the docs/content.zh/docs/dev/table/types.md to ensure that this page is consistent with the English version.
  -  Translate the new CAST documentation to Chinese.

## Verifying this change

Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluster with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no) no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no) no
  - The serializers: (yes / no / don't know) no
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know) no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know) no
  - The S3 file system connector: (yes / no / don't know) no

## Documentation

  - Does this pull request introduce a new feature? (yes / no) no
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by ChengkaiYang2022** [comment]
Hi, @RocMarshal ,cound you help me to review this PR? :-)

File: docs/content.zh/docs/dev/table/types.md [hunk 1]
```diff
@@ -191,6 +155,36 @@ For the Python Table API, those types are available in `pyflink.table.types.Data
...
+| `BOOLEAN`        |                                                    |
+| `BINARY`         |                                                    |
+| `VARBINARY`      |                                                    |
+| `BYTES`          |                                                    |
+| `DECIMAL`        | Supports fixed precision and scale.                |
+| `TINYINT`        |                                                    |
+| `SMALLINT`       |                                                    |
+| `INTEGER`        |                                                    |
+| `BIGINT`         |                                                    |
+| `FLOAT`          |                                                    |
```
Comment (ChengkaiYang2022) [inline docs/content.zh/docs/dev/table/types.md:174]: Adjust some lines in the docs/content.zh/docs/dev/table/types.md to ensure that this page is consistent with the English version. Please see [here](https://github.com/apache/flink/pull/18813/commits/da1c918a08dfcb50e27ffa0b54b36c4e711b460d)

**Comment by ChengkaiYang2022** [comment]
Hi,@RocMarshal , could you help to review this PR?:-)

**Comment by ChengkaiYang2022** [comment]
Hi,@fsk119 would you mind help to review this?

**Comment by fsk119** [comment]
Sorry for the late response. I will review it soon..

**Comment by fsk119** [comment]
Could you share the screen snapshot about the doc you modified?

**Comment by ChengkaiYang2022** [comment]
> Could you share the screen snapshot about the doc you modified?

Sure,Thanks for  reply!

The section I translate is the Casting part:

Chinese Version:




屏幕截图 2022-04-30 202719
屏幕截图 2022-04-30 202756

And the English version is:
屏幕截图 2022-04-30 202844


屏幕截图 2022-04-30 203023

The url is http://localhost:1313/flink/flink-docs-master/zh/docs/dev/table/types/#cast-%E6%96%B9%E6%B3%95

File: docs/content.zh/docs/dev/table/types.md [hunk 2]
```diff
@@ -1487,6 +1481,93 @@ Not supported.
 {{< /tab >}}
 {{< /tabs >}}
 
+<a name="casting"></a>
+
+cast 方法
+-------
+
+Flink Table API 和 Flink SQL 支持从 `输入` 数据类型 到 `目标` 数据类型的转换。有的转换
```
Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1489]: Remove '从' ？

File: docs/content.zh/docs/dev/table/types.md [hunk 3]
```diff
@@ -1487,6 +1481,93 @@ Not supported.
...
+无论输入值是什么都能保证转换成功，而有些转换则会发生错误，在运行时失败（即不可能转换为 `目标` 数据类型对应的值）。
```
Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1490]: > 而有些转换则会发生错误，在运行时失败

而有些转换则会在运行时失败

Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1490]: > 而有些转换则会发生错误，在运行时失败
而有些转换则会在运行时失败(例如，当无法为目标类型创建相应的值时)

File: docs/content.zh/docs/dev/table/types.md [hunk 4]
```diff
@@ -1487,6 +1481,93 @@ Not supported.
...
+例如，将 `INT` 数据类型的值转换为 `STRING` 数据类型一定能转换成功，但反过来从 `STRING` 数据类型
+转换为 `INT` 数据类型就不一定了。
```
Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1492]: > 但反过来从 `STRING` 数据类型转换为 `INT` 数据类型就不一定了。

但无法保证将`STRING` 数据类型转换为 `INT` 数据类型。

Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1492]: > 但反过来从 `STRING` 数据类型转换为 `INT` 数据类型就不一定了。

但无法保证将`STRING` 数据类型转换为 `INT` 数据类型。

File: docs/content.zh/docs/dev/table/types.md [hunk 5]
```diff
@@ -1487,6 +1481,93 @@ Not supported.
...
+
+在生成执行计划时，Flink 的 SQL 检查器会拒绝提交那些不可能直接转换为 `目标` 数据类型的SQL，并抛出 `ValidationException` 异常，
+例如想要从 `TIMESTAMP` 数据类型转化到 `INTERVAL` 数据类型。
```
Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1495]: 例如想要从 `TIMESTAMP` 数据类型转化到 `INTERVAL` 数据类型 -> 例如从 `TIMESTAMP` 类型转化到 `INTERVAL` 类型。

Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1495]: ??

File: docs/content.zh/docs/dev/table/types.md [hunk 6]
```diff
@@ -1487,6 +1481,93 @@ Not supported.
...
+然而有些查询即使通过了 SQL 检查器的验证，依旧可能会在运行期间转换失败，这就需要用户手动处理这些失败了。
```
Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1496]: 需要用户手动处理 -> 需要用户正确处理

Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1496]: 需要用户手动处理 -> 需要用户正确处理

File: docs/content.zh/docs/dev/table/types.md [hunk 7]
```diff
@@ -1487,6 +1481,93 @@ Not supported.
 {{< /tab >}}
 {{< /tabs >}}
 
+<a name="casting"></a>
+
+cast 方法
```
Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1486]: Upper case?

或者 翻译称为 类型转换？ 

In the PG community, they use the `类型转换`[1].

[1] http://www.postgres.cn/docs/12/typeconv.html

Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1486]: Upper case?

File: docs/content.zh/docs/dev/table/types.md [hunk 8]
```diff
@@ -1487,6 +1481,93 @@ Not supported.
...
+
+TRY_CAST('42' AS INT) --- 结果返回数字 42 的 INT 格式
+TRY_CAST(NULL AS VARCHAR) --- 结果返回 VARCHAR 类型的空值
+TRY_CAST('non-number' AS INT) --- 结果返回 INT 类型的空值
+COALESCE(TRY_CAST('non-number' AS INT), 0) --- 结果返回数字 0 的 INT 格式（非空）
+```
+
+下表展示了各个类型的转换程度，"Y" 表示支持，"!" 表示转换可能会失败，"N" 表示不支持：
+
+| 输入类型\输出类型                              | `CHAR`¹/<br/>`VARCHAR`¹/<br/>`STRING` | `BINARY`¹/<br/>`VARBINARY`¹/<br/>`BYTES` | `BOOLEAN` | `DECIMAL` | `TINYINT` | `SMALLINT` | `INTEGER` | `BIGINT` | `FLOAT` | `DOUBLE` | `DATE` | `TIME` | `TIMESTAMP` | `TIMESTAMP_LTZ` | `INTERVAL` | `ARRAY` | `MULTISET` | `MAP` | `ROW` | `STRUCTURED` | `RAW` |
```
Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1518]: 输出类型 -> 目标类型？

Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1518]: 输出类型 -> 目标类型？

File: docs/content.zh/docs/dev/table/types.md [hunk 9]
```diff
@@ -1487,6 +1481,93 @@ Not supported.
...
+| `ROW`                                  |                   Y                   |                    N                     |     N     |     N     |     N     |     N      |     N     |    N     |    N    |    N     |   N    |   N    |      N      |        N        |     N      |    N    |     N      |   N   |  !³   |      N       |   N   |
+| `STRUCTURED`                           |                   Y                   |                    N                     |     N     |     N     |     N     |     N      |     N     |    N     |    N    |    N     |   N    |   N    |      N      |        N        |     N      |    N    |     N      |   N   |   N   |      !³      |   N   |
+| `RAW`                                  |                   Y                   |                    !                     |     N     |     N     |     N     |     N      |     N     |    N     |    N    |    N     |   N    |   N    |      N      |        N        |     N      |    N    |     N      |   N   |   N   |      N       |  Y⁴   |
+
+备注：
+
+1. 在转换时，输入数据中的常量、变量会根据目标数据类型的格式进行修剪（trim）或填充（pad）。
```
Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1544]: 所有转化到具有固长或变长的类型时会根据类型的定义来裁剪或填充数据。

Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1544]: 所有转化到具有固长或变长的类型时会根据类型的定义来裁剪或填充数据。

File: docs/content.zh/docs/dev/table/types.md [hunk 10]
```diff
@@ -1487,6 +1481,93 @@ Not supported.
...
+2. 使用 `TO_TIMESTAMP` 方法和 `TO_TIMESTAMP_LTZ` 方法的场景，不要使用 `CAST` 或 `TRY_CAST`。
+3. 支持转换，当且仅当用其内部数据结构也支持转化时。转换可能会失败，当且仅当用其内部数据结构也可能会转换失败。
+4. 支持转换，当且仅当用使用 `RAW` 的类和类的序列化器一样。
+5. 支持转换，当且仅当用使用 `INTERVAL` 做“月”到“年”的转换。
+6. 支持转换，当且仅当用使用 `INTERVAL` 做“天”到“时间”的转换。
+
+无论是 `CAST` 还是 `TRY_CAST`，输入为 `NULL` ，则输出也为 `NULL`。
```
Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1551]: 请注意：无论是 `CAST` 还是 `TRY_CAST`，当输入为 `NULL` ，输出也为 `NULL`。

Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1551]: 请注意：无论是 `CAST` 还是 `TRY_CAST`，当输入为 `NULL` ，输出也为 `NULL`。

File: docs/content.zh/docs/dev/table/types.md [hunk 11]
```diff
@@ -1487,6 +1481,93 @@ Not supported.
...
+
+<a name="legacy-casting"></a>
+
+### 旧版本 cast 方法的兼容
+
+想使用 Flink 1.15 之前版本的 cast 方法，可以将参数 `table.exec.legacy-cast-behaviour` 设置为 `enabled`，
```
Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1557]: 用户可以通过将参数 `table.exec.legacy-cast-behaviour` 设置为 `enabled` 来启用 1.15 版本之前的 CAST 行为。

Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1557]: 用户可以通过将参数 `table.exec.legacy-cast-behaviour` 设置为 `enabled` 来启用 1.15 版本之前的 CAST 行为。

File: docs/content.zh/docs/dev/table/types.md [hunk 12]
```diff
@@ -1487,6 +1481,93 @@ Not supported.
...
+在 Flink 1.15 版本此参数默认为 disabled。
+
+如果设置为 enabled，请注意以下问题：
+
+* 转换为 `CHAR`/`VARCHAR`/`BINARY`/`VARBINARY` 数据类型时，不再自动修剪（trim）或填充（pad）。
+* 使用 `CAST` 时不再会因为转化失败而停止作业，只会返回 `NULL`，但不会像 `TRY_CAST` 那样推断正确的类型。
+* `CHAR`/`VARCHAR`/`STRING` 的转换结果会有一些细微的差别。
+
+{{< hint warning >}}
+我们 **不建议** 配置此参数，而是 **强烈建议** 在新项目中保持这个参数为默认禁用，以使用最新版本的 cast 方法。
```
Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1567]: 以使用最新版本的 cast 方法 -> 以使用最新版本的 cast 行为

Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1567]: 以使用最新版本的 cast 方法 -> 以使用最新版本的 cast 行为

File: docs/content.zh/docs/dev/table/types.md [hunk 13]
```diff
@@ -1487,6 +1481,93 @@ Not supported.
...
+例如，将 `INT` 数据类型的值转换为 `STRING` 数据类型一定能转换成功，但反过来从 `STRING` 数据类型
+转换为 `INT` 数据类型就不一定了。
+
+在生成执行计划时，Flink 的 SQL 检查器会拒绝提交那些不可能直接转换为 `目标` 数据类型的SQL，并抛出 `ValidationException` 异常，
+例如想要从 `TIMESTAMP` 数据类型转化到 `INTERVAL` 数据类型。
+然而有些查询即使通过了 SQL 检查器的验证，依旧可能会在运行期间转换失败，这就需要用户手动处理这些失败了。
+
+在 Flink Table API 和 Flink SQL 中，可以用下面两个内置方法来进行转换操作：
+
+* `CAST`：符合标准 SQL 的常规 cast 方法。在某些容易发生转换失败的查询场景中，当实际输入数据不是其 `输入` 数据类型时，作业便会运行失败。另外，类型校验机制支持输入的数据中有空值。
```
Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1500]: 符合标准 SQL 的常规 cast 方法 -> 定义在 SQL 标准的 CAST 方法。

当实际输入数据不是其 `输入` 数据类型时 -> 当实际输入数据不合法时。

> The type inference will preserve the nullability of the input type. 

类型推导会保留输入类型的可空性。也就是说，该方法的返回类型和输入类型有相同的可空性。

Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1500]: 符合标准 SQL 的常规 cast 方法 -> 定义在 SQL 标准的 CAST 方法。

当实际输入数据不是其 `输入` 数据类型时 -> 当实际输入数据不合法时。

> The type inference will preserve the nullability of the input type. 

类型推导会保留输入类型的可空性。

File: docs/content.zh/docs/dev/table/types.md [hunk 14]
```diff
@@ -1487,6 +1481,93 @@ Not supported.
...
+3. 支持转换，当且仅当用其内部数据结构也支持转化时。转换可能会失败，当且仅当用其内部数据结构也可能会转换失败。
+4. 支持转换，当且仅当用使用 `RAW` 的类和类的序列化器一样。
+5. 支持转换，当且仅当用使用 `INTERVAL` 做“月”到“年”的转换。
+6. 支持转换，当且仅当用使用 `INTERVAL` 做“天”到“时间”的转换。
+
+无论是 `CAST` 还是 `TRY_CAST`，输入为 `NULL` ，则输出也为 `NULL`。
+
+<a name="legacy-casting"></a>
+
+### 旧版本 cast 方法的兼容
```
Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1555]: 去掉 的兼容 吧

Comment (fsk119) [inline docs/content.zh/docs/dev/table/types.md:1555]: 去掉 的兼容 吧

**Review by fsk119 — COMMENTED** [review]
Thanks for your update. I left some comments.

File: docs/content.zh/docs/dev/table/types.md (hunk 7 above)
Comment (ChengkaiYang2022) [reply docs/content.zh/docs/dev/table/types.md:1486]: OK, maybe upper case 'CAST 方法' ? I think CAST will be just fine.

**Comment by ChengkaiYang2022** [comment]
Hi,@fsk119 ,thanks for your review and comments. I took all the suggestions and push a new [commit](https://github.com/apache/flink/pull/19498/commits/6750c63b53e14b1bc425942c065356187a85a40b), now the article reads much smoother now! Thanks, please take a look!

**Comment by ChengkaiYang2022** [comment]
屏幕截图 2022-05-06 170553
屏幕截图 2022-05-06 170607

**Review by fsk119 — APPROVED** [review]
Thanks for your update. LGTM.

**Comment by fsk119** [comment]
The CI test fails. Could you rebase to the latest master and trigger the test again?

**Comment by ChengkaiYang2022** [comment]
> The CI test fails. Could you rebase to the latest master and trigger the test again?

Okay,I will try to rebase

**Comment by ChengkaiYang2022** [comment]
@flinkbot

**Comment by ChengkaiYang2022** [comment]
@flinkbot run azure

**Review by RocMarshal — APPROVED** [review]
Thanks @ChengkaiYang2022 for the contribution and @fsk119 for the review.
LGTM , too.

**Comment by fsk119** [comment]
@flinkbot run azure

**Comment by ChengkaiYang2022** [comment]
> @flinkbot run azure

https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=35475&view=results

**Comment by ChengkaiYang2022** [comment]
> > @flinkbot run azure
> 
> https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=35475&view=results

@fsk119 hi,not sure why this failed?Is this because some azure resource problem?

**Comment by ChengkaiYang2022** [comment]
> > > @flinkbot run azure
> > 
> > 
> > https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=35475&view=results
> 
> @fsk119 hi,not sure why this failed?Is this because some azure resource problem?
屏幕截图 2022-05-09 114948

**Comment by ChengkaiYang2022** [comment]
The CI test faild because of some 'Test - finegrained_resource_management'? Not sure what that means.
Should I rebase from master again?

**Comment by fsk119** [comment]
@ChengkaiYang2022 I think we should make a try. Seems the CI is not stable..

**Comment by ChengkaiYang2022** [comment]
> @ChengkaiYang2022 I think we should make a try. Seems the CI is not stable..

Okay I will rebase again ASAP

**Comment by ChengkaiYang2022** [comment]
@flinkbot run azure

**Comment by ChengkaiYang2022** [comment]
CI test passed.[link](https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=35507&view=results)

**Review by leonardBang — APPROVED** [review]
Thanks @ChengkaiYang2022 and @fsk119 for the contribution and review, LGTM

**Comment by leonardBang** [comment]
@ChengkaiYang2022 Could you also open a backport PR for branch `release-1.15` ?

**Comment by ChengkaiYang2022** [comment]
> @ChengkaiYang2022 Could you also open a backport PR for branch `release-1.15` ?

Okay I will try to that.

**Pull request merged**